src/graph_iterator.rs

```diff
@@ -250,4 +186,58 @@ impl<R: ::std::io::BufRead, RF: Fn() -> R> EdgeMapper for ReaderMapper<R, RF> {
             }
         }
     }
-}
\ No newline at end of file
+}
+
+pub struct CachingReaderMapper<B: ::std::io::BufRead, F: Fn() -> B>  {
+    reader: ReaderMapper<B, F>,
+    upper:  Cell<Vec<((u16,u16), u32)>>,
+    lower:  Cell<Vec<(u16, u16)>>,
+    is_cached: Cell<bool>
+}
+
+impl <B: ::std::io::BufRead, F: Fn() -> B> CachingReaderMapper<B, F> {
+    pub fn new(reader: F, cap_upper: usize, cap_lower: usize) -> CachingReaderMapper<B, F> {
+        let mut upper = Vec::new();
+        upper.reserve(cap_upper);
+        let mut lower = Vec::new();
+        lower.reserve(cap_lower);
+        CachingReaderMapper::<B, F> {
+            reader: ReaderMapper::new(reader),
+            upper: Cell::new(upper),
+            lower: Cell::new(lower),
+            is_cached: Cell::new(false),
+        }
+    }
+}
+
+impl <B: ::std::io::BufRead, F: Fn() -> B> EdgeMapper for CachingReaderMapper<B, F> {
+    fn map_edges(&self, mut action: impl FnMut(u32, u32)) {
+        let mut upper = self.upper.take();
```

Comment:
This might not be zero-cost if `Vec::default()` does anything nontrivial. Might be better to use a `RefCell` for runtime borrow checking and leave the value in-place.

Reply:
Vec::default() doesn't do anything.